logic.py

sebastian2443 opened the pull request:
Ich habe mal unseren Code aus den Issues zusammengefasst.

sebastian2443 committed:
main.py

Ich habe mal unseren Code aus den Issues zusammengefasst.

seblin commented:
Irgendwie scheinen wir eine unterschiedliche Vorstellung von player.make_move() zu haben. Also meine Idee ist jedenfalls, dass in make_move() der Spieler aufgefordert wird, den nächsten Zug auf dem übergebenen Board durchzuführen. Dafür ist in Abhängigkeit vom verwendeten Frontend eine Abfrage auf der Shell oder in der GUI denkbar (später evtl auch als Berechnung, wenn der Spieler ein Computergegner ist). Da wir diese Abfrage beim derzeitigen Stand noch nicht machen, hatte ich Zufallswerte gewählt. Der hier eingereichte Pull Request verfälscht diese Idee total, da der Spielzug schon außerhalb erzeugt wird (weiß das Spiel etwa, was der Spieler will?) und der Spieler dient mehr oder weniger als "Durchgangsstation". Das finde ich nicht sehr objektorientiert. Wie sehen die anderen das?

EDIT:
Insgesamt finde ich es auch etwas unglücklich, dass hier nachträglich die API verändert wird. Dass der owner jetzt plötzlich ein player ist, kann man ja noch okay finden. Aber zusätzlich wird bei player.make_move() ein weiteres Argument ohne Rücksprache eingeführt und somit - wie oben angesprochen - der Methode eine ganz andere Bedeutung verliehen. Dann hätten wir uns dann UML-Diagramm auch sparen können, wenn man sich am Ende doch nicht dran hält...

EDIT2:
Auch get_allowed_fields() ist seltsam programmiert. Es werden erst leere Listen erstellt, die kurz darauf in Zeile 60 befüllt werden. Was soll das mit den eckigen Klammern für ein komisches Konstrukt sein? Eine Art Wegwerf-List-Comprehsion? Warum sollten wir diese nicht sofort an der Stelle einsetzen, wo jetzt quasi als Dummy eine leere Liste erzeugt wird?

sebastian2443 commented:
Um zu testen, ob die Methoden is_in_mill und get_allowed_fields funktionieren, habe ich erst einmal die
Feld Generierung aus dem make_move herausgenommen. Mit nur einem Feld lässt sich halt gar nichts testen. Das muss natürlich später geändert werden, sobald der dafür zuständige Code da ist.
Wir können das gerne auch direkt ändern, damit wir uns exakt ans UML halten.

Du hattest in deiner Methode is_in_mill den 'player' eingeführt. Ich fand die Idee gut, weil ich es sprechender als owner finde. Deswegen habe ich, anstelle deinen player zu owner zu ändern, alles auf player geswitched. Auch das können wir gerne wieder zu 'owner' umändern. 

Wegen deinem zweiten EDIT, ich programmiere das für gewöhnlich aus. Comprehensions sind noch sehr ungewohnt für mich. Ich brauche da also Übung. Ich habs jetzt auch kürzer hinbekommen.

sebastian2443 committed:
Update main.py

1. Player.make_move () wieder zum Ursprung von seblin
2. Field.player wegen dem UML zu Field.owner geändert
3. Den Kauderwelch aus get_allowed_fields bereinigt.

sebastian2443 committed:
Update main.py

sebastian2443 committed:
Update main.py

Ich habe in get_allowed_fields() die Variable candidates und die nachfolgende Comprehension angepasst.

seblin committed:
Implement first parts of the game logic

seblin commented:
Hab mal richtige Docstrings erstellt und die Datei in logic.py umbenannt. Jetzt bräuchten wir noch Tests. Und ja, es ist richtig dass man Zufallswerte schwierig testen kann. Hier ist es besser, direkt Werte vorzugeben. Ich finde, das sollte wir dann aber in player.make_move() machen und nicht außerhalb des Players.

sebastian2443 committed:
Unittest für logic.py 

Ich habe mal zwei Testszenarien erstellt, einmal für get_allowed_fields und für is_in_mill.
Damit der Code funktioniert, muss die Klasse Field aus logic.py um eine __eq__ Funktion erweitert werden.

sebastian2443 committed:
Update logic.py __eq__ Methode für Field

sebastian2443 commented:
Habe gestern eine __eq__ Methode für unsere Field Klasse hinzugefügt. Die habe ich für den Unittest gebraucht. Habe erst mal nur zwei Methoden auf richtigkeit geprüft. Ist der Unittest denn soweit 'Ok'?

ghost commented:
Ich kann da leider nicht sehr viel beisteuern, bin momentan auch etwas am strugglen weil im UML bspw. "init_fields()" als Methode definiert wurde, was im Code aber nur der normale __init__-Konstruktor ist, oder stellt man das so dar? 

Ich muss mich auch noch etwas mehr mit der Spiellogik beschäftigen, weil ich noch nicht sehe, warum es unbedingt notwendig ist, dass man eine Methode hat die alle möglichen Felder zum Setzen eines Steins anzeigt. Ich dachte mir ursprünglich, es sei einfacher wenn das in move_piece geprüft wird ob denn das Zielfeld bereits einen Owner besitzt etc.

seblin committed:
Verbesserte Tests

- Verwendung von geeigneten Methoden des Test-Frameworks
- Vermeidung einiger komplexer Konstrukte
- Aufruf von unittest.main() zum Starten

seblin committed:
Field.__eq__ entfernt

Dieses Verhalten kann ein namedtuple bereits von sich aus.

seblin committed:
owner beim Test von erlaubten Feldern entfernt

seblin commented:
@sls89:
Du meinst init_fields()? Die konnte ja mit einer einzigen Dict Comprehension erledigt werden. Daher haben wir sie nicht als eigene Methode geschrieben. Fändet du es besser, hierfür eine Methode zu haben?

Die get_allowed_fields() soll halt später auch zur visuellen Darstellung von möglichen Feldern dienen. Das ist zumindest meine Idee dahinter. Außerdem kann move_piece() diese Methode leicht abfragen, wenn es die Gültigkeit eines Spielzugs überprüft.

seblin committed:
SyntaxError behoben

Ein "in" war doppelt

ghost commented:
@seblin:
Nein, passt so. Ich war nur verunsichert weil die Methode ursprünglich laut UML anders hieß.

seblin commented:
Hab jetzt einfach mal das, was wir bisher hier gesammelt haben, als Commit hochgeladen. Ist ja alles etwas eingeschlafen hier...

sebastian2443 commented:
Wir waren auch leider wieder nur zu zweit unterwegs. Hatte gehofft, dass sls89 nach einer gewissen Einarbeitung den Einstieg findet. 
Dann wird es wohl wieder ein Winterschlaf..

seblin commented:
Jepp, jeder hat plötzlich etwas total wichtiges zu tun, wenn es ernst wird. Und das anscheinend ganzjährig. Ist schon etwas frustierend, aber naja...
